File: solution/09xxx/9012 - 괄호/9012.cpp
#include <iostream>
#include <string>
#include <stack>
using namespace std;

#define FASTIO cin.tie(NULL); cout.tie(NULL); ios::sync_with_stdio(false);

int main() {
    FASTIO

    int tc;

    cin >> tc;

    for(int testcase = 0; testcase < tc; testcase++) {
        string str;
        stack<char> st;
        bool stop_point = true;
        
        cin >> str;

        for(char strs : str) {
            if(strs == '(') st.push('(');
            else {
                if(st.size() == 0) {
                    stop_point = false;
                    break;
                }
                else st.pop();
            }
        }

        if(stop_point == true && st.size() == 0) cout << "YES" << "\n";
        else cout << "NO" << "\n";
    }
}

File: solution/09xxx/9012 - 괄호/9012.py
import sys
input = sys.stdin.readline

for _ in range(int(input())):
    data = input()[:-1]
    stack = list()
    result = "YES"

    for d in data:
        if d == '(':
            stack.append('(')
        else:
            if len(stack) == 0:
                result = "NO"
                break
            else:
                stack.pop()
    
    print(result if len(stack) == 0 else "NO")